en/services/command.md

```diff
@@ -80,12 +80,9 @@ sequenceDiagram;
 The rate at which progress messages are emitted is system-dependent.
 Generally though, the GCS should have a much increased timeout after receiving an ACK with `MAV_RESULT_IN_PROGRESS`.
 
-Typically if a system receives exactly the same command while it is already processing the command, it should continue the operation.
-If the command is a "resend" (confirmation number > 0) the command can be ignored; otherwise the system should respond immediately with a progress update `COMMAND_ACK`.
-
-If a system receives the same command with *different parameters* then in general it must be treated as *new command* (although this should be evaluated based on the specific operation requested):
-So for example:
-- If the new command cancels an operation, the recipient would stop the operation and emit `COMMAND_ACK` with `result=MAV_RESULT_ACCEPTED`
-  > **Note** If another (different) command was used to cancel the operation then it would be ACKd with `MAV_RESULT_ACCEPTED`, and the original command would be ACKd with `MAV_RESULT_FAILED`.
-- If the new command re-starts an operation the recipient should restart the job and emit `result=MAV_RESULT_IN_PROGRESS` etc.
-- If a new command changes some general value (e.g. a target position) this would be a new command.
+Often long running operations may be cancelled, restarted, or updated (continue operation but with new parameters) by sending the same command with different parameters.
+- A cancel command will complete immediately with response `COMMAND_ACK.result=MAV_RESULT_ACCEPTED`.
+- A restart or update command will start off a new sequence of progress updates - i.e. `COMMAND_ACK.result=MAV_RESULT_IN_PROGRESS` followed by updates until the operation completes.
```

Comment:
What's a restart or update command? Sorry, now I'm getting more and more confused.

Reply:
OK, the context here is the message that Nuno put in for updating firmware from a companion. The operation can be cancelled or restarted. The command has a parameter that can start, cancel or restart a firmware update. 

From that, it is clear to me that we need to:
-  clarify what a system should do if it gets (any) other command while processing a long running command
- clarify what a system should if it gets the _same_ command, either with the same or different parameters
- Explain recommended way to cancel. 

So a restart is "start the operation again". It isn't generally useful except for this case. You might reasonably suggest that you don't need this - receiving the start command twice would trigger a restart.

An update operation is when you want to send the same command but you want to update the parameters. So for example imagine I have a goto waypoint command or takeoff command that updates on progress and then sends accept when it reaches the waypoint [this is what I would do]. The command should be cancellable ('stop where you are"). However you might want to change where the end point is - so you send an update command - the same command but with a new lat/lon for the new position. 

IMO in this case the command should simply be treated as a new command - it should either be accepted, fail, or start a progress sequence.

Reply:
Restarting the update doesn't make much sense to me. I suggest we remove that. If something is stuck you can cancel it and start new but I don't see a scenario where you would want to restart something that you have already started.

> I have a goto waypoint command or takeoff command that updates on progress and then sends accept when it reaches the waypoint [this is what I would do]

The current implementation of goto acks immediately and then flies there, so it doesn't really apply there. If we were to change it to send in progress until it has reached the place, then your suggestion would make sense. However, commands should generally be used for something that you don't need to update, for things that you want to update and have a continuous stream of new setpoints, that's were other messages such as offboard position control, etc. come into place and commands are the wrong API.

This being said it's also tricky from the message flow. What if a long running command is updated? Do we then ack the initial command and also the update? Or do we deny the first one and accept the second one? And what if the update is invalid but the initial one was ok? Do we continue with the initial one and reject the updated one?
And how do we keep track if something is an update or just the next thing to do once we're done with the current command?

Reply:
Hi @julianoes 

I guess I don't like the accept-immediately approach we use for goto and takeoff because my main experience is writing scripts, and it is a pain to have to check telemetry to know when you're done. I guess we could abstract that in an SDK.

But yes, the message flow argument is compelling; that's pretty much why all the confusion. So yes, I think an operation can be cancelled. If this is done with the same command then the ACK is just cancelled, and is done once. If this was done with another command then the cancel command would ACK and the original command would complete with cancel. I'll triy this up next week.

Thanks for providing clarity.

Reply:
One more question @julianoes - what if the cancel command is sent when not in a long running operation - we say temp denied? Or do we just accept? 
Temp denied is more correct, but accept is more like "all good, move along", so your handling will be simpler.

Reply:
Why not also CANCELLED?

Reply:
Probably that is better - someone sent the message and they are waiting for that response. Even if it is "incorrect" the system should still end up being in the right place.